apps/src/calc/calc.js

```diff
@@ -247,135 +244,151 @@ function generateExpressionsFromBlockXml(blockXml) {
-
-    case 'functional_math_number':
-    case 'functional_math_number_dropdown':
-      var val = block.getTitleValue('NUM') || 0;
-      if (val === '???') {
-        val = 0;
-      }
-      return new Equation(null,
-        new ExpressionNode(parseInt(val, 10), [], block.id));
-
-    case 'functional_call':
-      name = block.getCallName();
-      var def = Blockly.Procedures.getDefinition(name, Blockly.mainBlockSpace);
-      if (def.isVariable()) {
-        return new Equation(null, new ExpressionNode(name));
-      } else {
-        var values = [];
-        var input, childBlock;
-        for (var i = 0; !!(input = block.getInput('ARG' + i)); i++) {
-          childBlock = input.connection.targetBlock();
-          // TODO - better default?
-          values.push(childBlock ? getEquationFromBlock(childBlock).expression :
-            new ExpressionNode(0));
-        }
-        return new Equation(null, new ExpressionNode(name, values));
-      }
-      break;
 
-    case 'functional_definition':
-      name = block.getTitleValue('NAME');
-      // TODO - access private
-      if (block.parameterNames_.length) {
-        name += '(' + block.parameterNames_.join(',') +')';
-      }
-      var expression = firstChild ? getEquationFromBlock(firstChild).expression :
-        new ExpressionNode(0);
+  // First evaluate both with the target set of inputs
+  if (targetSet.evaluateWithExpression(expression) !==
+      userSet.evaluateWithExpression(expression)) {
+    outcome.result = ResultType.FAILURE;
+    outcome.testResults = TestResults.LEVEL_INCOMPLETE_FAIL;
+    return outcome;
+  }
 
-      return new Equation(name, expression);
+  // At this point we passed using the target compute expression's inputs.
+  // Now we want to use all combinations of inputs in the range [-100...100],
+  // noting which set of inputs failed (if any)
+  var possibleValues = _.range(1, 101).concat(_.range(-0, -101, -1));
+  var numParams = expression.children.length;
+  var iterator = new InputIterator(possibleValues, numParams);
 
-    case 'functional_parameters_get':
-      return new Equation(null, new ExpressionNode(block.getTitleValue('VAR')));
+  var setChildToValue = function (val, index) {
+    // TODO - feels a little hacky directly modifying children
+    expression.children[index].value = val;
+  };
 
-    case 'functional_example':
-      // TODO (brent) - we dont do anything with functional_example yet, but
-      // this way we will at least persist it/not throw unknown type
-      return new Equation(null, null);
+  while (iterator.remaining() > 0 && !outcome.failedInput) {
+    var values = iterator.next();
+    values.forEach(setChildToValue);
 
-    default:
-      throw "Unknown block type: " + block.type;
+    if (targetSet.evaluateWithExpression(expression) !==
+        userSet.evaluateWithExpression(expression)) {
+      outcome.failedInput = values.slice();
```

Comment:
slightly prefer `_.clone(values)` for readability (does the same call under the covers)